src/agentscope/environment/env.py

```diff
@@ -0,0 +1,398 @@
+        """Remove a listener from the env.
+
+        Args:
+            target_event (`str`): The event function.
+            listener_name (`str`): The name of the listener to remove.
+
+        Returns:
+            `bool`: Whether the listener was removed successfully.
+        """
+
+    @abstractmethod
+    def get_listeners(self, target_event: str) -> List[EventListener]:
+        """Get the listeners of the specific event.
+
+        Args:
+            target_event (`str`): The event name.
+
+        Returns:
+            `List[EventListener]`: The listeners of the specific event.
+        """
+
+    @sync_func
+    @abstractmethod
+    def __getitem__(self, env_name: str) -> Env:
+        """Get a child env."""
+
+    @abstractmethod
+    def __setitem__(self, env_name: str, env: Env) -> None:
+        """Set a child env."""
+
+    @abstractmethod
+    def describe(self, **kwargs: Any) -> str:
+        """Describe the current state of the environment."""
+
+
+class BasicEnv(Env):
+    """A basic implementation of Env, which has no event function
+    and cannot get value.
+
+    Note:
+        `BasicEnv` is used as the base class to implement other
+        envs. Application developers should not use this class.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        listeners: dict[str, List[EventListener]] = None,
+        children: List[Env] = None,
+        parent: Env = None,
+    ) -> None:
+        """Init an BasicEnv instance.
+
+        Args:
+            name (`str`): The name of the env.
+            listeners (`dict[str, List[EventListener]]`, optional): The
+            listener dict. Defaults to None.
+            children (`List[Env]`, optional): A list of children
+            envs. Defaults to None.
+            parent (`Env`, optional): The parent env. Defaults
+            to None.
+        """
+        self._name = name
+        self.children = {
+            child.name: child for child in (children if children else [])
+        }
+        self.parent = parent
+        self.event_listeners = {}
+        if listeners:
+            for target_func, listener in listeners.items():
+                if isinstance(listener, EventListener):
+                    self.add_listener(target_func, listener)
+                else:
+                    for ls in listener:
+                        self.add_listener(target_func, ls)
+
+    @property
+    def name(self) -> str:
+        """Name of the env"""
+        return self._name
+
+    def get_parent(self) -> Env:
+        """Get the parent env of the current env.
+
+        Returns:
+            `Env`: The parent env.
+        """
+        return self.parent
+
+    def set_parent(self, parent: Env) -> None:
+        """Set the parent env of the current env.
+
+        Args:
+            parent (`Env`): The parent env.
+        """
+        if self.parent is not None:
+            self.parent.remove_child(self.name)
```

Comment:
why "remove_child" here?

Reply:
To keep the environments in a tree structure. Without `remove_child` here, the previous parent env can access the child node even after `set_parent`.

Reply:
aha, i was wrong about `parent` and  `self.parent`